binstat/binstat.go

```diff
@@ -0,0 +1,546 @@
+			whatLen = len(what)
+		}
+		//debug fmt.Printf("debug: detected ~%d%s in %s; using ~%d..\n", whatLenDefault, what[2:2 + whatLenDefault], what, whatLen)
+	}
+
+	p := BinStat{what[0:whatLen], t, funcName, fileLine, callerParams, callerTime, callerSize, callerSizeWhen}
+
+	t2 := runtimeNanoAsTimeDuration()
+
+	if verbose && global.verbose {
+		elapsedThisProc := time.Duration(t2 - global.startTimeMono).Seconds()
+		elapsedThisFunc := time.Duration(t2 - t).Seconds()
+		globalLog.Debug().Msgf("%f %d=pid %d=tid %s:%d(%s) // enter in %f // what[%s] .NumCPU()=%d .GOMAXPROCS(0)=%d .NumGoroutine()=%d",
+			elapsedThisProc, os.Getpid(), int64(C.gettid()), p.callerFunc, p.callerLine, p.callerParams, elapsedThisFunc, what, runtime.NumCPU(), runtime.GOMAXPROCS(0), runtime.NumGoroutine())
+	}
+
+	// for internal accounting, atomically increment counters in (never appended to) shadow array; saving additional lock
+	atomic.AddUint64(&global.frequencyShadow[internalEnter], 1)
+	atomic.AddUint64(&global.accumMonoShadow[internalEnter], uint64(t2-t))
+
+	return &p
+}
+
+func pointGeneric(p *BinStat, pointUnique string, callerSize int64, callerSizeWhen int, verbose bool) time.Duration {
+	// if binstat disabled, or <what> disabled; "~#" == "~<default len><what>"
+	if (!global.enable) || (2 == len(p.what)) {
+		return time.Duration(0)
+	}
+
+	t := runtimeNanoAsTimeDuration()
+
+	elapsedNanoAsTimeDuration := t - p.enterTime
+	if sizeAtLeave == callerSizeWhen {
+		p.callerSizeWhen = callerSizeWhen
+		p.callerSize = callerSize
+	}
+	var keySizeRange string
+	var keyTimeRange string
+	var pointType string
+	switch pointUnique {
+	case "Leave":
+		pointType = "leave"
+		switch p.callerSizeWhen {
+		case sizeAtEnter:
+			keySizeRange = fmt.Sprintf("/size[enter:%s]", x_2_y(float64(p.callerSize), true))
+		case sizeNotUsed:
+			keySizeRange = ""
+		case sizeAtLeave:
+			keySizeRange = fmt.Sprintf("/size[leave:%s]", x_2_y(float64(p.callerSize), true))
+		}
+	case "":
+	default:
+		pointType = "point"
+		keySizeRange = fmt.Sprintf("/size[point:%s]", pointUnique)
+	}
+	if p.callerTime {
+		elapsedSeconds := elapsedNanoAsTimeDuration.Seconds()
+		keyTimeRange = fmt.Sprintf("/time[%s]", x_2_y(elapsedSeconds, false))
+	}
+	key := fmt.Sprintf("/GOMAXPROCS=%d,CPUS=%d/what[%s]%s%s", runtime.GOMAXPROCS(0), runtime.NumCPU(), p.what, keySizeRange, keyTimeRange)
+
+tryAgainRaceCondition:
+	var frequency uint64
+	var accumMono uint64
+	globalRWMutex.RLock() // lock for many readers v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v v
+	index, keyExists := global.key2index[key]
+	if keyExists {
+		frequency = atomic.AddUint64(&global.frequency[index], 1)
+		accumMono = atomic.AddUint64(&global.accumMono[index], uint64(elapsedNanoAsTimeDuration))
+	}
+	globalRWMutex.RUnlock() // unlock for many readers ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^
+
+	if keyExists {
+		// full thru
+	} else {
```

Comment:
@turbolent thanks for the comment! It's personal taste but I find those small `!` modifiers easy to overlook, and therefore the code harder to read. But I made the change because there is only one line like that so far, and I don't want to draw out this PR :-)

I wonder if the linter could be made to auto detect code like this and auto flag it? That would save you having to comment :-)